### ISSUE TITLE
Support mapping on Handle.POST/PUSH

### ISSUE THREAD

**Issue by malx122** [Starcounter/Starcounter#3914]
Perfect for a "messaging" system between apps (right now we (mis)use GET for this which feels wrong and the mapping also requires answer from own app which makes it ugly and easy to miss for failure)

I have found many applications where this type of messaging is convenient and in some cases even required.

Example:
App1:
```
 Self.POST(UriMapping.MappingUriPrefix + "/app1_did_something_other_apps_might_want_to_respond_to/" + "whatIdid");
```
App2
```
Handle.POST("/app_2/handle_app1/{?}", (Request req, String data) =>
{
  //Do something
});

UriMapping.Map("/app_2/handle_app1/@w", UriMapping.MappingUriPrefix + " "/app1_did_something_other_apps_might_want_to_respond_to/@w");
```

**Comment by malx122** [comment]
Yes! I know that this functionality is possible in some ways with db hooks instead. However, there are use cases where db hooks just aren't a good/working solution.

**Comment by dan31** [comment]
Really interesting discussion. What we see is essentially https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern do on URI mapping API.
@warpech @joozek78 @miyconst opinions? Good, bad, ugly?

**Comment by mtornwall** [comment]
I recognize the need for transactional pubsub functionality in Starcounter, but I don't think HTTP handlers is the way to go. If something isn't *meant* to be available to external clients, it shouldn't be. What's to stop people from issuing malicious requests to these endpoints?

**Comment by miyconst** [comment]
I would like to review and refactor our current partial mapping and remove this kind of operations. It does not make sense to send any kind of requests GET or POST to communicate between apps on the same server.

The application communications should be done through database, the partials blending on a page should be done through `Starcounter.Partial.Retrieve`, and the application entry points should be defined with `Handle.GET`, `Handle.POST`.

**Comment by dan31** [comment]
@miyconst notice that the only reason @malx122 used it is inability to use commit hooks for that (meaning communicating through the database). Would you then recommend to strengthen the commit hooks in order to let the pattern be available in Starcounter? Provide your recommendation for @malx122 pattern.

**Comment by miyconst** [comment]
>However, there are use cases where db hooks just aren't a good/working solution.

@malx122, please provide some real examples when two apps should communicate and commit hooks can't achieve that.

**Comment by malx122** [comment]
First of all because polymorphic hooks is not supported, leading to unexpected behaviour. You can never create an db object in a hook because you don't know if any other apps have a hook on that specific db object.

The other is when you want user input on a specific use case:

* App 1 brings up a modal dialog "Create a new AAA!". Creating AAA is a quite complex thing to do and also depends on a lot of App1 options so creating AAA can not be done before the user actually press OK. So App 1 has no real Data model to work with, just a lot of options.

* App 1 does not now if any other apps want to have options to this modal dialog and so it still asks for partials. 

* App 2 now brings some options to the view, but still with no Data model so App 2 can not change db data depending on how the user chooses. 

* User presses OK and app 1 now creates AAA depending on the App 1 options. App 2 have no way of responding to this.

Option 1, Hooks on AAA creation:
Might work but only if the session can be retrieved in a hook and App 2 can get the result from its options in the create modal. However, this can give terrible side affects if AAA is created somewhere else in the code.
(Apart from the polymorphic hook problem above)

Option 2, Message:
App 1 posts a message that a AAA has been created with this modal (with the AAA db id), App 2 maps into this message and gets the session automatically and gets it options that it saved when giving the partial. No problem for app 2 to apply those options.

There is an option 3 where you can have a db object with just the options, but it will still give several problems that is to much to write here.

**Comment by miyconst** [comment]
>First of all because hook chains is not supported, leading to unexpected behaviour. You can never create an db object in a hook because you don't know if any other apps have a hook on that specific db object.

Is it possible to use a scheduled task to get the job done in your case?

>So App 1 has no real Data model to work with, just a lot of options.

It sounds a bit like a thick client, which we are trying to avoid.

>There is an option 3 where you can have a db object with just the options, but it will still give several problems that is to much to write here.

I would consider to use this approach. The idea is that the apps should not know about each other's existence.

Making a mapping of HTTP POST requests would definitely relate the apps between each other - one app expects another app to send a POST request to it, which is basically what REST does.

You can implement that on your own, just be aware, that serializing object for a POST request, and then deserializing it might be expensive. Better to share data between apps using database.

**Comment by malx122** [comment]
Yes, we could write apps that does not know of each other at all. That way we can forget anything useful for the end user which you realize after have tried to get a full app suite to work, it's just not possible. We are doing the next best thing. Apps that can both work together and independently.

My App 2 does not depend on App 1 in any way but can react to it.

The problem is that you are only testing a few small apps together, not building a system.

**Comment by miyconst** [comment]
>The problem is that you are only testing a few small apps together, not building a system.

Agree.

The sad fact is that:

- The database mapping is not yet implemented at all, only some prototype features are ready :(.
- The partials blending part is a subject of review/rework in the next Starcounter releases.

We will definitely take in account your requests and wishes when designing blending in Starcounter 3.0 or reworking existing solution.

For now adding another difficult to explain and maintain layer with POST request mappers would require unjustified effort.

**Comment by malx122** [comment]
I don't agree that it would ve unjustified or difficult to explain. For me it was a surprise that mapping did not work for POST, I tested it and expected it to work.

And just to add to the earlier. You are testing a few small apps and still breaking your own rules. All your apps are aware of Launcher and Signin even more so. That is not a complain, it is to point out that it is impossible to do get everything together if apps are not allowed to be aware of each other. That apps can both work together and independent is a lot better goal and a way to map (and remap) messages between apps is really important! If its done by url mapping or some other way is another question.

Option 3 is like I wrote not an option. Hooks are not allowed for creating/modifying db in apps as long as hook chains are not supported. ScheduleTask might work but gives other problems since you do not know when it is runned. And it is also a hack that is quite ugly/hard to explain.

**Comment by miyconst** [comment]
Okay, @malx122, I understand your need for a simpler and nicer way of communication between apps. Whether it will be HTTP POST request mapping or something else, is up to @dan31 to decide.

My personal opinion is that POST request mapping should not be implemented, but again, it is up to @dan31 to decide. Actually we are building a big app suite, which consists of almost 20 apps, and did not have a need for it.
